### PR TITLE
[tt-train] sdpa_bw: use 2-arg pack_reconfig_data_format to avoid SFPU drain

### PR DESCRIPTION
**Pull request by vmelnykovTT**
### Summary
Switch eight 1-arg `pack_reconfig_data_format(cb)` calls in the SDPA backward
kernels (Q and KV) to the conditional 2-arg form
`pack_reconfig_data_format(prev_cb, new_cb)`.
The 1-arg variant unconditionally reprograms the PACK destination format. On
Wormhole this triggers an implicit SFPU pipeline drain — the packer waits
for in-flight SFPU ops to retire before reconfiguring. Profiling attributed
~3.5 µs per K-iteration to this stall, previously hidden behind a generic
"pack-reconfig" zone and misread as PACK-bound work.
The 2-arg form takes `(old_cb, new_cb)` and the LLK compares
`pack_dst_format[old]` vs `pack_dst_format[new]`; when they match it skips
the reprogram (and the implicit drain). We pass the actual previous PACK
target so the call self-documents the data flow and so any future change to
a CB's static format is correctly handled by the LLK comparison instead of
being silently always-skipped.
### What's changed
`sdpa_bw_q_compute_kernel.cpp`:
- softmax-recompute pack: `cb_grad_query_accum → cb_attention_weights`

`sdpa_bw_compute_utils.hpp`:
- `compute_u_scalar_row` final pack: `cb_u_scalar_row → cb_u_scaler_output`
- `transpose_tile_fpu`: `cb_input → cb_transpose_wh` — KV K-loop hot path; `cb_input` equals the prev PACK target in both callers (`update_grad_value`, `update_grad_key`).
- `compute_grad_attn_weights`: `cb_prev_pack → cb_grad_attn_weights` — added `cb_prev_pack` parameter; Q passes `cb_attention_weights`, KV passes `cb_grad_value_accum`.
- `compute_grad_scores`: `cb_grad_attn_weights → cb_grad_scores`
- `update_grad_query`: `cb_grad_scores → cb_grad_query_accum`
- `update_grad_value`: `cb_transpose_wh → cb_grad_value_accum`
- `update_grad_key`: `cb_transpose_wh → cb_grad_key_accum`

### Impact
End-to-end TinyLlama training, 1000 steps, single-device WH N300, B=1, char dataset:
| metric           | baseline | this PR    | Δ                             |
|------------------|---------------------------------|------------|-------------------------------|
| loss @ step 1000 | 2.703125                        | 2.703125   | bit-identical (1000/1000 steps) |
| median step time | 1805.48 ms                      | 1786.72 ms | **−18.76 ms / −1.04%**        |

### CI Status
_Auto-generated on every push. Badges update live. Click a badge to filter runs by this branch._

- [](https://github.com/tenstorrent/tt-metal/actions/workflows/sanity-tests.yaml?query=branch:vmelnykov/sdpa_bw_pack_reconfig_fix)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/runtime-sanity-tests.yaml?query=branch:vmelnykov/sdpa_bw_pack_reconfig_fix)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml?query=branch:vmelnykov/sdpa_bw_pack_reconfig_fix)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/tt-metal-l2-nightly.yaml?query=branch:vmelnykov/sdpa_bw_pack_reconfig_fix)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/all-model-tests.yaml?query=branch:vmelnykov/sdpa_bw_pack_reconfig_fix)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select.yaml?query=branch:vmelnykov/sdpa_bw_pack_reconfig_fix)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-t3k.yaml?query=branch:vmelnykov/sdpa_bw_pack_reconfig_fix)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-galaxy.yaml?query=branch:vmelnykov/sdpa_bw_pack_reconfig_fix)

### CODE REVIEW

**Review by zbaczewskiTT — APPROVED** [review]
Thank you!

Does this impact Blackhole at all?

**Comment by vmelnykovTT** [comment]
@zbaczewskiTT 
I didn't check changes on BH VM

File: tt-train/sources/ttml/metal/ops/sdpa_bw/device/kernels/compute/sdpa_bw_q_compute_kernel.cpp [hunk 1]
```diff
@@ -161,12 +161,14 @@ FORCE_INLINE void process_single_row(uint32_t global_row_idx) {
         tile_regs_commit();
         tile_regs_wait();
         cb_reserve_back(cb_attention_weights, onetile);
-        pack_reconfig_data_format(cb_attention_weights);
+        // 2-arg pack_reconfig: skip reprogram if new PACK format matches the previously-configured one.
+        pack_reconfig_data_format(cb_grad_query_accum, cb_attention_weights);
```
Comment (mdragulaTT) [inline tt-train/sources/ttml/metal/ops/sdpa_bw/device/kernels/compute/sdpa_bw_q_compute_kernel.cpp:165]: Small question: after `compute_u_scalar_row()`, isn’t the previous pack target `cb_u_scaler_output` rather than `cb_grad_query_accum`? Since these CBs are all Float32 today this should be safe, but should we pass the actual previous pack CB here to keep the invariant clear?

**Review by mdragulaTT — APPROVED** [review]
LGTM overall. Before merge, please make sure tt-train Sanity and Nightly are green for this branch.

The WH TinyLlama result looks good, but limited. Not a blocker for this PR, but it would be useful to eventually have a small repeatable SDPA BW benchmark across a few shapes / batch sizes, similar to what we created with Ziemowit for other ops I think.

File: tt-train/sources/ttml/metal/ops/sdpa_bw/device/kernels/compute/sdpa_bw_q_compute_kernel.cpp (hunk 1 above)
Comment (vmelnykovTT) [reply tt-train/sources/ttml/metal/ops/sdpa_bw/device/kernels/compute/sdpa_bw_q_compute_kernel.cpp:165]: No, this part is a bit tricky.
On the first loop iteration, the previous pack target is cb_u_scaler_output, but after that it changes to cb_grad_query_accum, which is the last pack target used inside the loop.
I know my choice here is a little confusing, but adding extra conditions felt a bit overcomplicated.
Since I know that on the first loop iteration the previous pack data type already matches the current one, I skip the reconfiguration there.
